docs/example_navigation_ros_codegen.md

```diff
@@ -0,0 +1,527 @@
+
+To generate a ROS package you can use opengen - OpEn's Python interface (with opengen version `0.5.0` or newer). You will have to provide certain configuration parameters, such as the package name, the node name and the rate of your node in Hz.
+
+### Install opengen 0.5.0 in a virtual environment
+
+```bash
+cd ~
+mkdir open_ros_codegen
+cd ~/open_ros_codegen
+virtualenv -p python3.6 venvopen
+source venvopen/bin/activate
+pip install opengen==0.5.0a1
+pip install matplotlib
+```
+
+### Run the example code generation program
+
+Let's make a folder for the code generation script
+
+```bash
+cd ~/open_ros_codegen
+mkdir nmpc_open
+cd nmpc_open
+```
+
+Create a file named `create_open_solver.py` inside the `nmpc_open` folder with the following content
+
+```python
+import casadi as cs
+import opengen as og
+import matplotlib.pyplot as plt
+import numpy as np
+
+useSX = True  # Use CasADi SX or MX. Seems quicker with SX
+doSim = False  # Run simulation after creating the solver
+
+N = 10
+NX = 3
+NU = 2
+sampling_time = 0.1
+NSim = 100
+
+Q = cs.DM.eye(NX) * [1.0, 1.0, 0.0001]
+R = cs.DM.eye(NU) * [0.1, 50.0]
+QN = cs.DM.eye(NX) * [10.0, 10.0, 0.0001]
+
+
+def dynamics_ct(_x, _u):
+    return cs.vcat([_u[0] * cs.cos(_x[2]),
+                    _u[0] * cs.sin(_x[2]),
+                    _u[1]])
+
+
+def dynamics_dt(_x, _u):
+    dx = dynamics_ct(_x, _u)
+    return cs.vcat([_x[i] + sampling_time * dx[i] for i in range(NX)])
+
+
+def stage_cost(_x, _u, _x_ref=None, _u_ref=None):
+    if _x_ref is None:
+        _x_ref = cs.DM.zeros(_x.shape)
+    if _u_ref is None:
+        _u_ref = cs.DM.zeros(_u.shape)
+    dx = _x - _x_ref
+    du = _u - _u_ref
+    return cs.mtimes([dx.T, Q, dx]) + cs.mtimes([du.T, R, du])
+
+
+def terminal_cost(_x, _x_ref=None):
+    if _x_ref is None:
+        _x_ref = cs.DM.zeros(_x.shape)
+    dx = _x - _x_ref
+    return cs.mtimes([dx.T, QN, dx])
+
+if useSX:
+    x_0 = cs.SX.sym("x_0", NX)
+    x_ref = cs.SX.sym("x_ref", NX)
+    u_k = [cs.SX.sym('u_' + str(i), NU) for i in range(N)]
+else:
+    x_0 = cs.MX.sym("x_0", NX)
+    x_ref = cs.MX.sym("x_ref", NX)
+    u_k = [cs.MX.sym('u_' + str(i), NU) for i in range(N)]
+
+
+x_t = x_0
+total_cost = 0
+
+for t in range(0, N):
+    total_cost += stage_cost(x_t, u_k[t], x_ref)  # update cost
+    x_t = dynamics_dt(x_t, u_k[t])  # update state
+
+total_cost += terminal_cost(x_t, x_ref)  # terminal cost
+
+optimization_variables = []
+optimization_parameters = []
+
+optimization_variables += u_k
+optimization_parameters += [x_0]
+optimization_parameters += [x_ref]
+
+optimization_variables = cs.vertcat(*optimization_variables)
+optimization_parameters = cs.vertcat(*optimization_parameters)
+
+umin = [-2.0, -1.0] * N  # - cs.DM.ones(NU * N) * cs.inf
+umax = [2.0, 1.0] * N  # cs.DM.ones(NU * N) * cs.inf
+
+bounds = og.constraints.Rectangle(umin, umax)
+
+problem = og.builder.Problem(optimization_variables, optimization_parameters, total_cost).with_constraints(bounds)
+
+if doSim:
```

Comment:
I would drop `doSim` as well. By the way, if you have compiled an optimizer once, the second time you re-compile it takes significantly less time.